platforms/core-configuration/file-collections/src/main/java/org/gradle/api/internal/file/collections/DefaultConfigurableFileCollection.java

```diff
@@ -211,7 +212,8 @@ public boolean startVisit(FileCollectionInternal.Source source, FileCollectionIn
                 if (DefaultConfigurableFileCollection.this == fileCollection) {
                     throw new UnsupportedOperationException("Self-referencing ConfigurableFileCollections are not supported. Use the from() method to add to a ConfigurableFileCollection.");
                 }
-                return true;
+                // Configurations cannot be a result of (a + b)
+                return !(fileCollection instanceof Configuration);
```

Comment:
:x: This is problematic for two reasons:

1) it links to `Configuration` that is a higher-level concept that should not be visible from `:file-collections`, and is visible only because of historic reasons.
2) the reason we are skipping digging deeper here is not because this is a `Configuration` specifically, but because we don't want to trigger resolving things eagerly for this check. This reason is not documented here.

I think a better approach would be to introduce something like `ResolvableFileCollection`, let `Configuration` implement it and check for that here. Doing so would both document the purpose of the check, and break the dependency on a the higher-level concept of `Configuration`.

Reply:
You could even add something like `ResolvableFileCollection.hasAlreadyBeenResolved()` and call it from here. That way we could even detect recursive references when they are embedded in `Configuration` objects. Not sure if it's necessary, though, to weed out simple `a = a + b` calls, and I am leaning to not doing it. Just putting all on the table.

Reply:
To me this would be a good middle ground to not reach out of `:file-collections`. @lptr Are you ok with such solution?

@wolfs Could you cherry-pick this change to be based on master, so we can merge it there directly? If you have other work I can take care of that

Reply:
It would be great if you could take care of it.